php/libraries/SubprojectController.class.inc

```diff
@@ -0,0 +1,252 @@
+<?php
+/**
+ * This file contains the SubprojectController class.
+ *
+ * PHP Version 7
+ *
+ * @category Controller
+ * @package  Subprojects
+ * @author   Rida Abou-Haidar <rida.abou-haidar@mcin.ca>
+ * @license  http://www.gnu.org/licenses/gpl-3.0.txt GPLv3
+ * @link     https://www.github.com/aces/Loris/
+ */
+
+namespace LORIS;
+
+/**
+ * Base class that contains method to access the subproject of projects
+ *
+ * @category Controller
+ * @package  Subprojects
+ * @author   Rida Abou-Haidar <rida.abou-haidar@mcin.ca>
+ * @license  http://www.gnu.org/licenses/gpl-3.0.txt GPLv3
+ * @link     https://www.github.com/aces/Loris/
+ */
+class SubprojectController
+{
+    /**
+     * Array holding existing instances
+     *
+     * @var array
+     */
+    private static $_instances = [];
+
+    /**
+     * The database connection to use
+     *
+     * @var \Database
+     */
+    protected $database;
+
+    /**
+     * The constructor
+     *
+     * @param \Database $database The database connection to use
+     */
+    public function __construct(\Database $database)
+    {
+        if (!$database->isConnected()) {
+            throw new \RuntimeException(
+                "SubprojectController - Database not connected"
+            );
+        }
+        $this->database = $database;
+    }
+
+    /**
+     * Retruns a list of configured Subprojects
+     *
+     * @return array The list of subprojects
+     */
+    public function getAllSubprojects(): array
+    {
+        $subprojectRows = $this->database->pselect(
+            "SELECT
+                SubprojectID,
+                title,
+                useEDC,
+                WindowDifference,
+                RecruitmentTarget
+            FROM subproject
+            ",
+            []
+        );
+
+        $subprojects = [];
+        foreach ($subprojectRows as $row) {
+            $subproject    = new \LORIS\Subproject(
+                intval($row['SubprojectID']),
+                $row['title'],
+                $row['useEDC'] === '1',
+                $row['WindowDifference'],
+                isset($row['RecruitmentTarget'])
+                    ? intval($row['RecruitmentTarget']) : null
+            );
+            $subprojects[] = $subproject;
```

Comment:
The instantiation can be done from the PDO using `PDO::FETCH_CLASS`. I think it would be an elegant way to reduce code duplication. The same validations and type conversions are done here and in `getSubprojectByTitle`.

What would need to be done though is to have a constructor without parameters and to do the type conversion inside it.

Reply:
Henri had implemented, `objectToSQL()` and `sqlToObject()` functions to do this works. you wanna go with that ?

Reply:
I don't think we need any function to do it. What I am saying is that PDO can handle this.

Reply:
Basically, 
```
public function getAllSubprojects(): \Traversable
{
    $statement = $this->_db->prepare($selectstring);
    $statement->setFetchMode(\PDO::FETCH_CLASS, '\LORIS\Subproject');
    yield from $statement->execute();
}

public function getSubprojectByID(string $id): Subproject
{
    $statement = $this->_db->prepare($selectstring . 'WHERE SubprojectID = :v_id');
    $statement->setFetchMode(\PDO::FETCH_CLASS, '\LORIS\Subproject');
    $statement->execute(['v_id' => $id]);
    return $statement->fetch();
}
```

Reply:
if that's going to become a common idiom I would think a $DB->pselectInto  and pselectOneInto would be a better idea than duplicate code everywhere.. I can send a PR, but what's the point of `yield from` when execute already returns a \Traversable?

Reply:
@xlecours how does PDO map what column maps to what variable ?

Reply:
@driusan you are right, `yield from` is redondant. And you could add `$DB->pselectInto` but you will have to alter the execute function.

Reply:
@ridz1208 It sets the object variable as the column names from your query. Then it call the constructor.

Reply:
I don't think I like that solution. 

1. I'm trying to dissociate the object from the DB as much as possible and keep the DAO/Controller more DB involved
2. It seems to instantiate the variables implicitly and I don't like that
3. some sub-entities like projects or visits will anyways be loaded by the controller and will have to be defined separately making a mix of implicit and explicit variables...

All in all it's kind of cool but I think I would go for it if there was a way to define the variables explicitly and typed and then map them somehow to the DB column names using PDO. not the current implementations

Reply:
I understand you like writing code. It was just a suggestion.